agents/suprsync/suprsync.py

```diff
@@ -0,0 +1,152 @@
+    args : Namespace
+        Namespace with parsed arguments
+
+    Attributes
+    ------------
+    agent : OCSAgent
+        OCS agent object
+    log : txaio.tx.Logger
+        txaio logger object, created by the OCSAgent
+    archive_name : string
+        Name of the managed archive. Sets which files in the suprsync db should
+        be copied.
+    ssh_host : str, optional
+        Remote host to copy data to. If None, will copy data locally.
+    ssh_key : str, optional
+        ssh-key to use to access the ssh host.
+    remote_basedir : path
+        Base directory on the destination server to copy files to
+    db_path : path
+        Path of the sqlite db to monitor
+    delete_after : float
+        Seconds after which this agent will delete successfully copied files.
+    cmd_timeout : float
+        Time (sec) for which cmds run on the remote will timeout
+    copy_timeout : float
+        Time (sec) after which a copy command will timeout
+    timeout_wait : float
+        Time (sec) to sleep after a timeout occurs.
+    """
+    def __init__(self, agent, args):
+        self.agent = agent
+        self.log = txaio.make_logger()
+        self.archive_name = args.archive_name
+        self.ssh_host = args.ssh_host
+        self.ssh_key = args.ssh_key
+        self.remote_basedir = args.remote_basedir
+        self.db_path = args.db_path
+        self.delete_after = args.delete_after
+        self.max_copy_attempts = args.max_copy_attempts
+        self.running = False
+        self.cmd_timeout = args.cmd_timeout
+        self.copy_timeout = args.copy_timeout
+        self.timeout_wait = args.timeout_wait
+
+    def run(self, session, params=None):
+        """run()
+
+        **Process** - Main run process for the SupRsync agent. Continuosly
+        checks the suprsync db checking for files that need to be handled.
+        """
+
+        srfm = SupRsyncFilesManager(self.db_path, create_all=True)
+
+        self.running = True
+        session.set_status('running')
+
+        handler = SupRsyncFileHandler(
+            srfm, self.remote_basedir, delete_after=self.delete_after,
+            ssh_host=self.ssh_host, ssh_key=self.ssh_key,
+            cmd_timeout=self.cmd_timeout, copy_timeout=self.copy_timeout
+        )
+
+        while self.running:
+            with srfm.Session.begin() as session:
+                file = srfm.get_next_file(
+                    self.archive_name, session=session,
+                    delete_after=self.delete_after,
+                    max_copy_attempts=self.max_copy_attempts)
+                if file is not None:
+                    try:
+                        handler.handle_file(file, session)
+                    except subprocess.TimeoutExpired:
+                        self.log.error(
+                            "Timed out when processing {path}",
+                            path=file.local_path)
+                        time.sleep(self.timeout_wait)
+                time.sleep(3)
```

Comment:
Each iteration of this loop opens a session to the database every iteration, rsync's a single file, and takes 3-5+ seconds (3-5 seconds of sleep, plus operation time). This seems inefficient.

We touched up on this on the telecon today, but this should be reworked to handle more than one file at once. A good workflow I think would be to:
- Check the database for all files that need to be copied
- Assemble that list for passing to rsync
  - You could provide a long list of files at the command line, or write to a temporary file and use `--files-from`
  - The list will need to be built with paths relative to the remote-basedir, since that's the only thing specified on the other end, not the full path for each file like you're doing now.
  - This also avoids the whole need to create the remote location first for each subdirectory, as long as the remote-basedir exists (so you should still create that)
- rsync all files that need to be sent at once
- checksum remote files
  - I'm less sure of a way to do this efficiently remotely. Perhaps with the persistent ssh connection that Matthew suggested it won't be much of a problem
- update the database

You could run that loop once every few minutes or so.